src/zkevm_specs/pi_circuit.py

```diff
@@ -281,64 +313,87 @@ def verify_circuit(
     """
 
     rows = witness.rows
-    table = witness.calldata_gas_cost_table
-
-    # 1.0 rand_rpi copy constraint from public input to advice column
-    assert rows[0].rand_rpi == witness.public_inputs.rand_rpi
-
-    # 1.1 rpi_rlc copy constraint from public input to advice column
-    assert rows[0].rpi_rlc_acc == witness.public_inputs.rpi_rlc
-
-    # 1.2 chain_id copy constraint from public input to raw_public_inputs
-    assert rows[BlockTag.ChainId].raw_public_inputs == witness.public_inputs.chain_id
-
-    # 1.3 state_root copy constraint from public input to raw_public_inputs
-    offset_extra = BLOCK_LEN + 2
-    assert rows[offset_extra + 2].raw_public_inputs == witness.public_inputs.state_root.lo.expr()
-    assert rows[offset_extra + 3].raw_public_inputs == witness.public_inputs.state_root.hi.expr()
-
-    # 1.4 state_root_prev copy constraint from public input to raw_public_inputs
-    assert (
-        rows[offset_extra + 4].raw_public_inputs == witness.public_inputs.state_root_prev.lo.expr()
-    )
-    assert (
-        rows[offset_extra + 5].raw_public_inputs == witness.public_inputs.state_root_prev.hi.expr()
-    )
+    calldata_gas_cost_table = witness.calldata_gas_cost_table
+    public_inputs = witness.public_inputs
+    keccak_table = witness.keccak_table
+    block_table = witness.block_table
+    tx_table = witness.tx_table
+    copy_constrains = witness.copy_constrains
```

Comment:
I don't understand the modeling you're doing of the copy constraints.
This variable has type `List[List[int]]`.  But each copy constraint is a set of cells (each identified by a column and offset) that must be equal.  Could you explain how that is modeled?

I also don't understand why the copy_constraints appear in the Witness.  I would imagine that they would appear as assert equals in the `verify_circuit`.  Something like:
```python
for i in range(10):
  assert witness.column_a[OFFSET+i] == witness.column_b[i]
```

Reply:
Regarding to type `List[List[int]]`, the nested list will be aggregate to single cell below `bytes_to_fq`.

Here I applied a tricky optimisation on byte vertical layout. Since for block table and transaction table, some require hi + lo, whereas some just require lo, therefore there is no fixed offset concept here. In Halo2, there is a `assigned cell` object which represent this tricks easily. However in Python spec there is no `assigned cells` concept. therefore to simulate copy constrains, below `pop` the byte**s**  one by one and compare with table value.

I can move the `bytes_to_fq` to make it happened earilier before enter `verifying_circuit`, so it will be `List[int]` which more close to list of cell, wdyt

Reply:
> Regarding to type `List[List[int]]`, the nested list will be aggregate to single cell below via `bytes_to_fq`.
> 
> Here I applied a tricky optimisation on byte vertical layout. Since for block table and transaction table, there can be [1, 8, 20, 32] bytes. Hence to efficient represeted in vertical byte layout, offset are dymanic, and there is no fixed offset concept here. In Halo2, there is a `assigned cell` object which represent this tricks easily. However in Python spec there is no `assigned cells` concept. therefore to simulate copy constrains, below `pop` the byte**s** one by one and compare with table value.

I see, I think I follow the reasoning now.  So the `copy_constraints` variable contains the `rpi_bytes` already grouped by value entries from the tables right?  I still need to think more about it and review more to give more feedback later.

Also something that confused me is that `copy_constraints` is `List[List[int]]` but many times it's filled with `List[bytes]`.  Also `bytes_to_fq` takes `bytes` or `Sequence[FQ]` but not `List[int]`.  I'm not sure if there's a typing issue here or I'm missing something.  If I run `mypy .` I don't get any errors on `pi_circuit.py`.  But if I run `mypy src/zkevm_specs/pi_circuit.py` I get this:
```
src/zkevm_specs/pi_circuit.py:334: error: Argument 1 to "bytes_to_fq" has incompatible type "List[int]"; expected "bytes"  [arg-type]
src/zkevm_specs/pi_circuit.py:335: error: Argument 1 to "bytes_to_fq" has incompatible type "List[int]"; expected "bytes"  [arg-type]
src/zkevm_specs/pi_circuit.py:338: error: Argument 1 to "bytes_to_fq" has incompatible type "List[int]"; expected "bytes"  [arg-type]
src/zkevm_specs/pi_circuit.py:342: error: Argument 1 to "bytes_to_fq" has incompatible type "List[int]"; expected "bytes"  [arg-type]
src/zkevm_specs/pi_circuit.py:343: error: Argument 1 to "bytes_to_fq" has incompatible type "List[int]"; expected "bytes"  [arg-type]
src/zkevm_specs/pi_circuit.py:347: error: Argument 1 to "bytes_to_fq" has incompatible type "List[int]"; expected "bytes"  [arg-type]
src/zkevm_specs/pi_circuit.py:348: error: Argument 1 to "bytes_to_fq" has incompatible type "List[int]"; expected "bytes"  [arg-type]
src/zkevm_specs/pi_circuit.py:356: error: Argument 1 to "bytes_to_fq" has incompatible type "List[int]"; expected "bytes"  [arg-type]
src/zkevm_specs/pi_circuit.py:358: error: Argument 1 to "bytes_to_fq" has incompatible type "List[int]"; expected "bytes"  [arg-type]
src/zkevm_specs/pi_circuit.py:363: error: Argument 1 to "bytes_to_fq" has incompatible type "List[int]"; expected "bytes"  [arg-type]
src/zkevm_specs/pi_circuit.py:364: error: Argument 1 to "bytes_to_fq" has incompatible type "List[int]"; expected "bytes"  [arg-type]
src/zkevm_specs/pi_circuit.py:367: error: Argument 1 to "bytes_to_fq" has incompatible type "List[int]"; expected "bytes"  [arg-type]
src/zkevm_specs/pi_circuit.py:376: error: Argument 1 to "bytes_to_fq" has incompatible type "List[int]"; expected "bytes"  [arg-type]
src/zkevm_specs/pi_circuit.py:377: error: Argument 1 to "bytes_to_fq" has incompatible type "List[int]"; expected "bytes"  [arg-type]
src/zkevm_specs/pi_circuit.py:380: error: Argument 1 to "bytes_to_fq" has incompatible type "List[int]"; expected "bytes"  [arg-type]
src/zkevm_specs/pi_circuit.py:384: error: Incompatible types in assignment (expression has type "Row", variable has type "TxTableRow")  [assignment]
src/zkevm_specs/pi_circuit.py:462: error: Need type annotation for "raw_bytes_group" (hint: "raw_bytes_group: List[<type>] = ...")  [var-annotated]
src/zkevm_specs/pi_circuit.py:521: error: Incompatible default for argument "value_hi" (default has type "None", argument has type "bytes")  [assignment]
src/zkevm_specs/pi_circuit.py:521: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
src/zkevm_specs/pi_circuit.py:521: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
src/zkevm_specs/pi_circuit.py:523: error: Argument 1 to "append" of "list" has incompatible type "bytes"; expected "int"  [arg-type]
src/zkevm_specs/pi_circuit.py:524: error: Argument 1 to "append" of "list" has incompatible type "bytes"; expected "int"  [arg-type]
src/zkevm_specs/pi_circuit.py:525: error: Argument 1 to "append" of "list" has incompatible type "bytes"; expected "int"  [arg-type]
src/zkevm_specs/pi_circuit.py:527: error: Argument 1 to "append" of "list" has incompatible type "bytes"; expected "int"  [arg-type]
src/zkevm_specs/pi_circuit.py:583: error: Incompatible return value type (got "List[bytes]", expected "List[int]")  [return-value]
src/zkevm_specs/pi_circuit.py:598: error: Incompatible return value type (got "List[bytes]", expected "List[List[int]]")  [return-value]
src/zkevm_specs/pi_circuit.py:613: error: Incompatible return value type (got "List[bytes]", expected "List[List[int]]")  [return-value]
src/zkevm_specs/pi_circuit.py:722: error: Argument 1 to "append" of "list" has incompatible type "bytes"; expected "int"  [arg-type]
src/zkevm_specs/pi_circuit.py:723: error: Argument 1 to "append" of "list" has incompatible type "bytes"; expected "int"  [arg-type]
src/zkevm_specs/pi_circuit.py:725: error: Argument 1 to "append" of "list" has incompatible type "bytes"; expected "int"  [arg-type]
src/zkevm_specs/pi_circuit.py:726: error: Argument 1 to "append" of "list" has incompatible type "bytes"; expected "int"  [arg-type]
src/zkevm_specs/pi_circuit.py:727: error: Argument 1 to "flattern_len" has incompatible type "List[int]"; expected "List[List[Any]]"  [arg-type]
src/zkevm_specs/pi_circuit.py:737: error: Argument 1 to "extend" of "list" has incompatible type "List[List[int]]"; expected "Iterable[int]"  [arg-type]
src/zkevm_specs/pi_circuit.py:742: error: Argument 1 to "flattern_len" has incompatible type "List[int]"; expected "List[List[Any]]"  [arg-type]
src/zkevm_specs/pi_circuit.py:758: error: Argument 1 to "extend" of "list" has incompatible type "List[List[int]]"; expected "Iterable[int]"  [arg-type]
src/zkevm_specs/pi_circuit.py:772: error: Argument 1 to "flattern_len" has incompatible type "List[int]"; expected "List[List[Any]]"  [arg-type]
src/zkevm_specs/pi_circuit.py:782: error: No overload variant of "reversed" matches argument type "int"  [call-overload]
src/zkevm_specs/pi_circuit.py:782: note: Possible overload variants:
src/zkevm_specs/pi_circuit.py:782: note:     def [_T] reversed(self, Reversible[_T], /) -> reversed[_T]
src/zkevm_specs/pi_circuit.py:782: note:     def [_T] reversed(self, SupportsLenAndGetItem[_T], /) -> reversed[_T]
src/zkevm_specs/pi_circuit.py:875: error: Argument 17 to "Row" has incompatible type "KeccakTable"; expected "KeccakTableRow"  [arg-type]
src/zkevm_specs/pi_circuit.py:904: error: Argument "copy_constrains" to "Witness" has incompatible type "List[int]"; expected "List[List[int]]"  [arg-type]
Found 37 errors in 1 file (checked 1 source file)
```
This reports many instances of incompatibility between `List[int]` and `bytes` related to the `copy_constrains`, but there are also some other type issues.  I would like to know why `mypy .` doesn't report these issues.

> I can move the `bytes_to_fq` to make it happened earilier before enter `verifying_circuit`, so it will be `List[int]` which more close to list of cell, wdyt

Let me continue reviewing before answering this!

Reply:
Fixed all types issues in latest [commit](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/417/commits/70edc68143fd8ad9908a546f6c8402d9058c4451)

For `mypy .`  seems it doesn't work to check all py source. Try a workaround `mypy src/zkevm_specs/__init__.py` to fix this issue first

Reply:
Update on mypy: To make mypy check all source file, I try to fix it, but still failed to check all sources. Final work around is via bash `find` and pass exactly py files, then mypy can successfully capture all src files.

I raised a [issue](https://github.com/privacy-scaling-explorations/zkevm-specs/issues/429) to tracking this